functions/src/index.ts

```diff
@@ -114,13 +131,78 @@ export const dropPin = functions.https.onCall(
   }
 );
 
+/**
+ * calculateReward will give the first NUM_BONUS people an exponentially decaying bonus
+ * above BASE, while pin finders after NUM_BONUS will get a reward slightly under BASE
+ * SCALING_REDUCTION exists so that the first pin finder doesn't get like 10x BASE
+ * reward scaling function visualized:
+ * |
+ * |
+ * #
+ * |#
+ * | #
+ * |  #
+ * |   #
+ * |    #
+ * |     #
+ * |      #
+ * |       #
+ * |        #
+ * |          #
+ * |           ##
+ * |             ##
+ * |                ##
+ * |                  ###
+ * |                      ###
+ * |                          ####
+ * |                               #####
+ * |                                     ########
+ * |                                               ##########
+ * |                                                            ################
+ * |                                                                                  ################
+ * ___________________________________________________________________________________________________
+ */
+// async function calculateReward(pin: Pin, transaction?: firestore.Transaction) {
+//   const BASE = 100;
+//   const NUM_BONUS = 20;
+//   const SCALING_REDUCTION = 2;
+
+//   return (NUM_BONUS / (pin.finds + SCALING_REDUCTION)) * BASE;
+// }
+
+/**
+ * calculateCost finds the number of pins within three progressively larger radii
+ * and linearly scales the cost of a new pin according to number of neighbors
+ */
 async function calculateCost(
   latitude: number,
   longitude: number,
   transaction?: firestore.Transaction
 ) {
-  // TODO: implement algorithm for calculating pin cost
-  return 200;
+  const BASE = 60;
+  const CLOSE_RADIUS = 0.1;
+  const MID_RADIUS = 0.5;
+  const FAR_RADIUS = 2;
+  const CLOSE_SCALE = 1.0;
+  const MID_SCALE = 0.25;
+  const FAR_SCALE = 0.1;
+
+  const close = (await getPinsNearby(latitude, longitude, CLOSE_RADIUS)).length;
+  const mid =
+    (await getPinsNearby(latitude, longitude, MID_RADIUS)).length - close;
+  const far =
+    (await getPinsNearby(latitude, longitude, FAR_RADIUS)).length - mid - close;
```

Comment:
Could we do this more efficiently? Maybe only run the query for the `FAR_RADIUS` and have some helper function that calculates the distance for each pin within that radius?

Reply:
Most likely. I'm thinking Pin reward should be a function of cost, so I'll likely revisit both of these algorithms now that the underlying functionality is there.

Reply:
@ckleinhans I could add this in a separate PR too so that this one doesn't get stale, i.e. merge this base functionality in and then start on a fresh "improve algos" PR.

Reply:
Works for me! I'll resolve this and make a new task for that.